csrc/preseg_passes/fmin_fmax_promotion.cpp

```diff
@@ -0,0 +1,298 @@
+              status = IterDomainStatus::BAD_BROADCAST;
+            } else if (status != IterDomainStatus::NONE) {
+              status = IterDomainStatus::GOOD_BROADCAST;
+            }
+          }
+
+          iterMap[out_id] = status;
+        } else {
+          if (StatusIsBad(status)) {
+            return false;
+          }
+        }
+      }
+
+    } else if (expr->isA<BinaryOp>()) {
+      auto* left_tv = dynamic_cast<TensorView*>(expr->input(0));
+      auto* right_tv = dynamic_cast<TensorView*>(expr->input(1));
+
+      // One side (not both) might not be a TensorView.
+      // To handle this, just propagate the status of the other side.
+      if (!left_tv) {
+        left_tv = right_tv;
+      } else if (!right_tv) {
+        right_tv = left_tv;
+      }
+
+      auto left2right = logical_map.mapBestEffort(
+          left_tv->domain(),
+          left_tv->getLogicalDomain(),
+          right_tv->domain(),
+          right_tv->getLogicalDomain());
+
+      auto left2out = logical_map.mapBestEffort(
+          left_tv->domain(),
+          left_tv->getLogicalDomain(),
+          out_tv->domain(),
+          out_tv->getLogicalDomain());
+
+      for (IterDomain* left_id : left_tv->getLogicalDomain()) {
+        // Note: this assumes that the left <-> right mapping exists
+        // Does this need to handle left-right mapping failures?
+
+        IterDomainStatus leftStatus = iterMap[left_id];
+        IterDomainStatus rightStatus = iterMap[left2right[left_id]];
+
+        IterDomainStatus status = BopStatus(leftStatus, rightStatus);
+
+        auto out_id = left2out[left_id];
+
+        if (out_id) {
+          iterMap[out_id] = status;
+        } else {
+          if (StatusIsBad(status)) {
+            return false;
+          }
+        }
+      }
+
+    } else {
+      // unknown op type, ensure it has no bad status since information will not
+      // flow through it.
+      if (anyBadInputs) {
+        return false;
+      }
+    }
+  }
+
+  // Check whether any bad status reached output nodes
+  auto output_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+  for (TensorView* tv : output_tvs) {
+    for (IterDomain* id : tv->getLogicalDomain()) {
+      IterDomainStatus status = iterMap[id];
+      if (StatusIsBad(status)) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+void FMinFMaxPromotionPass::runPass(Fusion* fusion) {
+  FusionGuard fusion_guard(fusion);
+
+  // The outer loop runs over all expressions, filtering out most of them.
+  // It stops only on min/max reductions, which become the target for the rest
+  // of the analysis.
+  for (Expr* targetExpr : fusion->exprs()) {
+    auto* targetRop = dynamic_cast<ReductionOp*>(targetExpr);
+
+    if (!targetRop) {
+      continue;
+    }
+
+    auto reduction_type = targetRop->getReductionOpType();
+
+    if (reduction_type == BinaryOpType::Min ||
+        reduction_type == BinaryOpType::Max) {
+      if (AnalyzeMinMaxOp(targetRop)) {
```

Comment:
TODO: The current strategy should have a bug when downstream nodes are promoted. This may break the analysis from upstream nodes. A simple way to fix this should be:
- Run the analysis in reverse-topological order, and
- Make the analysis treat FMin/FMax ops as graph outputs.

Reply:
No, I should create one